embassy-rp/src/i2c_slave.rs

```diff
@@ -181,14 +181,28 @@ impl<'d, T: Instance> I2cSlave<'d, T> {
                         p.ic_rx_tl().write(|w| w.set_rx_tl(11));
                     }
 
+                    // rd_req is not cleared so that the clock is stretched into `respond_to_read`.
                     if stat.restart_det() && stat.rd_req() {
+                        p.ic_clr_restart_det().read();
                         Poll::Ready(Ok(Command::WriteRead(len)))
                     } else if stat.gen_call() && stat.stop_det() && len > 0 {
+                        p.ic_clr_gen_call().read();
+                        p.ic_clr_stop_det().read();
                         Poll::Ready(Ok(Command::GeneralCall(len)))
-                    } else if stat.stop_det() {
+                    } else if stat.stop_det() && len > 0 {
+                        p.ic_clr_stop_det().read();
                         Poll::Ready(Ok(Command::Write(len)))
                     } else if stat.rd_req() {
                         Poll::Ready(Ok(Command::Read))
+                    // If restart_det() is set but not rd_req() the register can
+                    // get stuck, and we can enter a tight loop.
+                    } else if stat.restart_det() && !stat.rd_req() {
+                        p.ic_clr_restart_det().read();
+                        Poll::Pending
+                    // Likewise if stop_det() is set and len is 0.
+                    } else if stat.stop_det() && p.ic_rxflr().read().rxflr() == 0 {
```

Comment:
I'm seeing this a lot with i2c across two distinct chips (it does not appear with the loopback). I'm not sure if I've done something wrong, or if it's a race condition.